src/msw/spinctrl.cpp

```diff
@@ -746,9 +769,18 @@ int wxSpinCtrl::GetOverlap() const
     return wxGetWindowRect(m_hwndBuddy).right - wxGetWindowRect(GetHwnd()).left;
 }
 
+wxSize wxSpinCtrl::GetBestSizeFromDigitsCount(int digitsCount) const
+{
+    const wxString largestString('8', digitsCount);
+    return DoGetSizeFromTextSize(GetTextExtent(largestString).GetWidth());
+}
+
 wxSize wxSpinCtrl::DoGetBestSize() const
 {
-    return DoGetSizeFromTextSize(FromDIP(DEFAULT_ITEM_WIDTH));
+    const int base = GetBase();
+    const int digitsCount = wxMax(GetDigitsCount(GetMin(), base),
+                                  GetDigitsCount(GetMax(), base));
+    return GetBestSizeFromDigitsCount(digitsCount);
```

Comment:
Why do we have to go through "digits count"? Couldn't we just use `GetSizeFromTextSize(wxMax(GetTextExtent(wxString::Format("%d", GetMin())).x, GetTextExtent(wxString::Format("%d", GetMax())).x))`?

Or do what `wxSpinCtrlGTKBase::DoGetBestSize()` does (of course, this should be extracted into a common function in `wxSpinCtrlBase` then).

Reply:
`wxSpinCtrl` under `MSW` is derived from `wxSpinButton` so the common parent for `wxSpinCtrlGTKBase` (`GTK`) and `wxSpinCtrl` (`MSW`) is `wxControl` but not `wxSpinCtrlBase`. So I get rid of math calculations and use string formatting locally in `wxSpinCtrl`.

Reply:
Yes, the proper solution for this would be to use CRTP, i.e. create `wxSpinCtrlIface<T>` deriving from `T` and redefine `wxSpinCtrlBase` as `wxSpinCtrlIface<wxControl>`, while inheriting `wxSpinCtrl` in wxMSW from `wxSpinCtrlIface<wxSpinButton>`, but it's probably not worth doing this right now.

If there is enough of code to reuse, you could easily do it in some inline function in `include/wx/private/spinctrl.h` however.